pkg/controller/datavolume-controller.go

```diff
@@ -2599,3 +2623,288 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
+func (r *DatavolumeReconciler) detectCloneSize(
+	dv *cdiv1.DataVolume,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	cloneType cloneStrategy) (bool, error) {
+
+	var targetSize resource.Quantity
+	sourcePvc, err := r.findSourcePvc(dv)
+	if err != nil {
+		return false, err
+	}
+	sourceSize := sourcePvc.Status.Capacity.Storage()
+
+	// Due to possible filesystem overhead complications when cloning
+	// using host-assisted strategy, we create a pod that automatically
+	// collects the size of the original virtual image with 'qemu-img'.
+	// If another strategy is used or the original PVC's volume mode
+	// is "block", we simply extract the value from the original PVC's spec.
+	if cloneType == HostAssistedClone && getVolumeMode(sourcePvc) == corev1.PersistentVolumeFilesystem {
+		// If available, we first try to get the virtual size from previous iterations
+		imgSize, available := getSizeFromAnnotations(sourcePvc)
+		if !available {
+			imgSize, err = r.getSizeFromPod(sourcePvc)
+			if err != nil {
+				return false, err
+			} else if imgSize == ImgSizeNotReady {
+				return false, nil
+			}
+		}
+		// We enable the clone-controller to skip the size comparison requirement
+		dv.Annotations[AnnPermissiveClone] = "true"
+		// If needed, inflate size with filesystem overhead
+		targetSize, err = inflateSizeWithOverhead(r.client, imgSize, pvcSpec, sourceSize)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		targetSize = *sourceSize
+	}
+
+	pvcSpec.Resources.Requests[corev1.ResourceStorage] = targetSize
+	return true, nil
+}
+
+// getSizeFromAnnotations checks the source PVC's annotations and returns the requested size if it has already been obtained
+func getSizeFromAnnotations(sourcePvc *corev1.PersistentVolumeClaim) (string, bool) {
+	virtualImageSize, available := sourcePvc.Annotations[AnnVirtualImageSize]
+	if available == true {
+		sourceCapacity, available := sourcePvc.Annotations[AnnSourceCapacity]
+		currCapacity := sourcePvc.Status.Capacity
+		// Checks if the original PVC's capacity has changed
+		if available && currCapacity.Storage().Cmp(resource.MustParse(sourceCapacity)) == 0 {
+			return virtualImageSize, true
+		}
+	}
+
+	return "0", false
+}
+
+// getSizeFromPod attempts to get the image size from a pod that directly obtains said value from the source PVC
+func (r *DatavolumeReconciler) getSizeFromPod(sourcePvc *corev1.PersistentVolumeClaim) (string, error) {
+	// The pod should not be created until the source PVC has finished the import process
+	if !hasPvcSucceeded(sourcePvc) {
+		return ImgSizeNotReady, nil
+	}
+
+	pod, err := r.createSizeDetectionPod(sourcePvc)
+	if err != nil {
+		return ImgSizeNotReady, err
+	} else if !hasPodSucceeded(pod) {
+		return ImgSizeNotReady, nil
+	}
+	// Parse raw image size from the pod's termination message
+	imgSize := pod.Status.ContainerStatuses[0].State.Terminated.Message
+	err = r.client.Delete(context.TODO(), pod)
```

Comment:
Might want to wait with deleting the pod until you are sure everything you wanted to save the value in has updated. Like the updateClonePVCAnnotations below. If that fails for whatever reason, now your pod is gone and you have to redo the entire pod creation cycle to get the value again. If you reverse the order, then you are sure the update was successful and you can delete confident the value is on the PVC.